design/ip-address-management-for-networkdata.md

```diff
@@ -0,0 +1,328 @@
+kind: Metal3IPPool
+metadata:
+  name: pool-1
+  namespace: default
+  ownerReferences:
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    kind: Metal3Cluster
+    name: cluster-1
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    kind: Metal3Data
+    name: metal3data-1
+spec:
+  pools:
+    - start: 192.168.0.10
+      end: 192.168.0.15
+      subnet: 192.168.0.0/24
+      step: 1
+    - start: 192.168.0.20
+      end: 192.168.0.25
+      subnet: 192.168.0.0/24
+status:
+  lastUpdated: "2020-04-02T06:36:09Z"
+  allocations:
+    "192.168.0.11": "metal3data-1"
+  addresses:
+    "metal3data-1": "pool-1-192-168-0-11"
+  error: false
+  errorMessage: ""
+```
+
+In the *spec* of *Metal3IPPool*, there would be a *pools* list, that would
+contain a list of IP address pools, with the *start* and *end* attributes
+giving the start and end ip addresses of the pool, the *subnet* allowing to
+verify that the allocated IP is in the pool and from which the start and end ip
+addresses can be inferred. The *step* item gives the interval between allocated
+IP addresses. Specifying single ip addresses can be achieved by setting the
+start and end ip address to that single ip address.
+
+The *status* would contain a *lastUpdated* field with the timestamp of the last
+update and and *error* boolean that is set to true if an error during allocation
+happens. In that case, the *errorMessage* will contain the text of the error.
+The *allocations* map will map the IP address to the *Metal3Data* object it was
+allocated for and the *addresses* will map the *metal3Data* objects with the
+*Metal3IPAddress* objects.
+
+All IP addresses specified in a Metal3IPPool must be from the same subnet.
+
+The *Metal3IPAddress* object would be the following
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3IPAddress
+metadata:
+  name: pool-1-192-168-0-11
+  namespace: default
+  ownerReferences:
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    kind: Metal3IPPool
+    name: pool-1
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    kind: Metal3Data
+    name: metal3data-1
+spec:
+  Metal3Data:
+    Name: metal3data-1
+  Address: 192.168.0.11
+  Metal3IPPool:
+    Name: pool-1
+status:
+  ready: true
+```
+
+For each owner reference added on a *Metal3IPPool* that is a *Metal3Data*, the
+controller reconciling the *Metal3IPPool* will select an available IP address.
+If all the *Metal3Data* owner references are from the same *Metal3DataTemplate*,
+then the controller will use the index of the *Metal3Data* to compute the IP
+address to allocate by sorting the addresses in incremental order and selecting
+the IP address corresponding to the index. In case of conflict, (multiple
+Metal3DataTemplate referencing the same *Metal3IPPool*, a random available IP
+address will be selected.
+
+Once the IP address is selected, the controller will create a *Metal3IPAddress*
```

Comment:
If the allocation is only stored in statuses, it will be lost across the pivoting operation. The allocation needs to be stored in a spec part of an object, so that the statuses can be recreated after pivoting. There were two alternatives, using an independent object, as is proposed here, or add a field in the spec of Metal3Data. the second option required the reconciler to fetch the allocations from the Metal3Data objects when recreating the status. It is more straightforward to use a separate object, as the reconciler then only relies on those objects to recreate the status. It also allows more flexibility in the API, since an external actor could also create a Metal3IPAddress to reserve an IP for example. Both approaches would work, this one being probably cleaner in the API design.

Reply:
We could also use the same trick as in BMO, storing a copy of the status in the annotation, however, I don't think this is a proper way to solve the issue. Using separate objects gives us additional benefits such as preventing conflict upon ip allocation, similarly to the index in Metal3DataTemplates.

Reply:
In addition, the deletion of the Metal3Data object would automatically trigger the deletion of the Metal3IPAddress, ensuring that the allocations are in sync, and there are no leftovers allocations.

Reply:
OK, it sounds like maintaining state across the pivot is the main use. We should state that explicitly in the doc as the reason for having both CRDs, because some of the rest of this section makes a lot more sense with that context. :-)

Reply:
ok, I'll expand it in the user stories and motivations